include/exadg/fluid_structure_interaction_precice/interface_coupling.h

```diff
@@ -0,0 +1,185 @@
+    : precice(precice)
+  {
+  }
+
+  void
+  setup(std::shared_ptr<const MatrixFree<dim, Number>> matrix_free_dst,
+        unsigned int const                             dof_index_dst,
+        std::vector<quad_index> const &                quad_indices_dst_in,
+        MapBoundaryCondition const &                   map_bc_in,
+        const std::string &                            mesh_name,
+        const std::string &                            data_name)
+  {
+    quad_rules_dst = quad_indices_dst_in;
+    map_bc         = map_bc_in;
+
+    for(auto quadrature : quad_rules_dst)
+    {
+      // initialize maps
+      map_index_dst.emplace(quadrature, MapIndex());
+      map_q_points_dst.emplace(quadrature, ArrayQuadraturePoints());
+      map_solution_dst.emplace(quadrature, ArrayTensor());
+
+      MapIndex &              map_index          = map_index_dst.find(quadrature)->second;
+      ArrayQuadraturePoints & array_q_points_dst = map_q_points_dst.find(quadrature)->second;
+      ArrayTensor &           array_solution_dst = map_solution_dst.find(quadrature)->second;
+
+
+      /*
+       * 1. Setup: create map "ID = {face, q, v} <-> vector_index" and fill array of quadrature
+       * points
+       */
+      for(unsigned int face = matrix_free_dst->n_inner_face_batches();
+          face <
+          matrix_free_dst->n_inner_face_batches() + matrix_free_dst->n_boundary_face_batches();
+          ++face)
+      {
+        // only consider relevant boundary IDs
+        if(map_bc.find(matrix_free_dst->get_boundary_id(face)) != map_bc.end())
+        {
+          Integrator integrator(*matrix_free_dst, true, dof_index_dst, quadrature);
+          integrator.reinit(face);
+
+          for(unsigned int q = 0; q < integrator.n_q_points; ++q)
+          {
+            Point<dim, VectorizedArray<Number>> q_points = integrator.quadrature_point(q);
+
+            for(unsigned int v = 0; v < VectorizedArray<Number>::size(); ++v)
+            {
+              Point<dim> q_point;
+              for(unsigned int d = 0; d < dim; ++d)
+                q_point[d] = q_points[d][v];
+
+              Id                      id    = std::make_tuple(face, q, v);
+              types::global_dof_index index = array_q_points_dst.size();
+              map_index.emplace(id, index);
+              array_q_points_dst.push_back(q_point);
+            }
+          }
+        }
+      }
+      array_solution_dst.resize(array_q_points_dst.size());
+    }
+
+    ArrayQuadraturePoints map_q_points_dst_precice;
+
+    for(auto quadrature : quad_rules_dst)
+    {
+      ArrayQuadraturePoints & array_q_points_dst = map_q_points_dst.find(quadrature)->second;
+
+      for(const auto i : array_q_points_dst)
+        map_q_points_dst_precice.emplace_back(i);
+    }
+
+    // finally, give boundary condition access to the data
+    for(auto boundary : map_bc)
+      boundary.second->set_data_pointer(map_index_dst, map_solution_dst);
```

Comment:
Ok, so you would propose to adapt the base class' C++ interface of the `Adapter::CcouplingInterface` in order to get to the following hierarchy: `Adapter::CouplingInterface` (with potentially a different name) and the derived classes `DoFInterface`, `QuadInterface` and the `ExaDG::InterfaceCoupling`. If not, we should maybe have an online meeting to discuss on this.

Reply:
I would not derive `ExaDG::InterfaceCoupling` from `Adapter::CouplingInterface`, since this would again mean that `InterfaceCoupling` depends on precice-related implementations. My point is that the responsibilities of `InterfaceCoupling` (used for precice-FSI) and `Adapter::Adapter` are not 100% clear to me and I wonder whether one could improve that. I have to add that I am currently also learning the code, so it might be that some comments are in contradiction to each other or that I have not understood the code well enough.

Here, I am also referring to this comment: https://github.com/exadg/exadg/pull/148#discussion_r798610140. Want I don't like so much is that some calls in the driver (write) are invoked by precice, while others (read) are invoked by communicator (which is of type `InterfaceCoupling`). I would prefer that things look more "uniform" in the driver, in the sense that those calls are all invoked by `precice` or `communicator`. I am not 100% sure which variant is better or whether this really makes sense, but do you understand my point?

Reply:
Ok, yes, I need to think about a proper solution for that.